File: .agents/scripts/update_wiki.ts
/**
 * ==============================================================================
 * 🌐 Gitea Wiki Synchronization and Deployment Service (update_wiki.ts)
 * ==============================================================================
 * @description  OpenKB로 정제된 concepts 및 summaries 문서들을
 *               gitea-wiki 저장소 및 Obsidian 보관소 경로로 복사하고 원격지에 push/동기화합니다.
 * @constraints  SSL 인증서 에러 방지를 위해 git http.sslVerify=false 구동.
 *               Strict Typing 및 OOP Patterns 아키텍처 규칙을 상시 준수합니다.
 * @dependencies Node.js runtime, fs, path, child_process
 * ==============================================================================
 */

import { execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

class WikiConfig {
  public readonly projectRoot: string = '/Users/ejpark/workspace/scraper';
  public readonly openkbDir: string;
  public readonly wikiDir: string;

  constructor() {
    this.openkbDir = path.join(this.projectRoot, 'data/openkb');
    this.wikiDir = path.join(this.projectRoot, 'data/gitea-wiki');
  }
}

class FileService {
  public copyDirSync(src: string, dest: string): void {
    fs.mkdirSync(dest, { recursive: true });
    const entries = fs.readdirSync(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = path.join(src, entry.name);
      const destPath = path.join(dest, entry.name);

      if (entry.isDirectory()) {
        this.copyDirSync(srcPath, destPath);
      } else {
        fs.copyFileSync(srcPath, destPath);
      }
    }
  }
}

class WikiSynchronizer {
  private config: WikiConfig;
  private files: FileService;

  constructor(config: WikiConfig, files: FileService) {
    this.config = config;
    this.files = files;
  }

  private runCmd(cmd: string, cwd?: string): void {
    try {
      execSync(cmd, { cwd, encoding: 'utf-8', stdio: 'inherit' });
    } catch (error) {
      const err = error as Error;
      console.error(`❌ [WikiSync] 명령어 실행 실패: ${cmd}`);
      console.error(err.message);
      process.exit(1);
    }
  }

  private runCmdOutput(cmd: string, cwd?: string): string {
    try {
      return execSync(cmd, { cwd, encoding: 'utf-8', stdio: ['ignore', 'pipe', 'pipe'] }).trim();
    } catch {
      return '';
    }
  }

  public sync(): void {
    console.log('🤖 Synchronizing compiled concepts & summaries to Gitea Wiki...');
    fs.mkdirSync(this.config.wikiDir, { recursive: true });

    const conceptsDir = path.join(this.config.openkbDir, 'wiki/concepts');
    if (fs.existsSync(conceptsDir)) {
      this.files.copyDirSync(conceptsDir, this.config.wikiDir);
    }

    const summariesDir = path.join(this.config.openkbDir, 'wiki/summaries');
    if (fs.existsSync(summariesDir)) {
      this.files.copyDirSync(summariesDir, path.join(this.config.wikiDir, 'summaries'));
    }

    // Git Push 동기화 진행
    console.log('📤 Pushing to Gitea Wiki remote...');
    this.runCmd('git config http.sslVerify false', this.config.wikiDir);

    const status = this.runCmdOutput('git status --porcelain', this.config.wikiDir);
    if (status) {
      this.runCmd('git add .', this.config.wikiDir);
      const dateStr = new Date().toISOString().replace(/T/, ' ').replace(/\..+/, '');
      this.runCmd(`git commit -m "chore: auto-compound agent session logs ${dateStr}"`, this.config.wikiDir);
      this.runCmd('git push origin master', this.config.wikiDir);
      console.log('✅ Gitea Wiki Remote synchronisation complete.');
    } else {
      console.log('   No changes detected in Wiki. Push skipped.');
    }

    console.log('🎉 Wiki Deployment and Synchronization Pipeline finished.');
  }
}

// Execute
const config = new WikiConfig();
const files = new FileService();
const synchronizer = new WikiSynchronizer(config, files);
synchronizer.sync();

File: apps/openkb/src/update_wiki.py
import os
import shutil
import subprocess
from datetime import datetime
from pathlib import Path

class WikiConfig:
    def __init__(self):
        if Path('/data').exists():
            self.project_root = Path('/data')
            self.openkb_dir = self.project_root / 'openkb'
            self.wiki_dir = self.project_root / 'gitea-wiki'
        else:
            self.project_root = Path('/Users/ejpark/workspace/scraper')
            self.openkb_dir = self.project_root / 'data/openkb'
            self.wiki_dir = self.project_root / 'data/gitea-wiki'

def copy_dir(src: Path, dest: Path):
    if not src.exists():
        return
    dest.mkdir(parents=True, exist_ok=True)
    for item in os.listdir(src):
        s = src / item
        d = dest / item
        if s.is_dir():
            copy_dir(s, d)
        else:
            shutil.copy2(s, d)

def run_cmd(cmd: str, cwd: Path):
    try:
        subprocess.run(cmd, shell=True, cwd=cwd, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except subprocess.CalledProcessError as e:
        print(f"❌ [WikiSync] 명령어 실행 실패: {cmd}")
        print(str(e))
        exit(1)

def run_cmd_output(cmd: str, cwd: Path) -> str:
    try:
        res = subprocess.run(cmd, shell=True, cwd=cwd, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        return res.stdout.strip()
    except subprocess.CalledProcessError:
        return ""

def main():
    print('🤖 Synchronizing compiled concepts & summaries to Gitea Wiki...')
    config = WikiConfig()
    os.chdir(config.openkb_dir)
    config.wiki_dir.mkdir(parents=True, exist_ok=True)

    concepts_dir = config.openkb_dir / 'wiki/concepts'
    if concepts_dir.exists():
        copy_dir(concepts_dir, config.wiki_dir)

    summaries_dir = config.openkb_dir / 'wiki/summaries'
    if summaries_dir.exists():
        copy_dir(summaries_dir, config.wiki_dir / 'summaries')

    # git repository인지 안전성 검사 추가
    git_dir = config.wiki_dir / '.git'
    if not git_dir.exists():
        print('⚠️ [WikiSync] gitea-wiki 디렉토리에 .git 저장소 정보가 존재하지 않습니다. Git 동기화를 스킵합니다.')
        print('🎉 Wiki Deployment and Synchronization Pipeline finished (Skipped Git Push).')
        return

    print('📤 Pushing to Gitea Wiki remote...')
    run_cmd('git config http.sslVerify false', config.wiki_dir)
    run_cmd('git config user.name "WikiSync Agent"', config.wiki_dir)
    run_cmd('git config user.email "agent@openkb.local"', config.wiki_dir)
    
    status = run_cmd_output('git status --porcelain', config.wiki_dir)
    if status:
        run_cmd('git add .', config.wiki_dir)
        date_str = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        commit_cmd = (
            'GIT_AUTHOR_NAME="WikiSync Agent" '
            'GIT_AUTHOR_EMAIL="agent@openkb.local" '
            'GIT_COMMITTER_NAME="WikiSync Agent" '
            'GIT_COMMITTER_EMAIL="agent@openkb.local" '
            f'git commit -m "chore: auto-compound agent session logs {date_str}"'
        )
        run_cmd(commit_cmd, config.wiki_dir)
        run_cmd('git push origin main', config.wiki_dir)
        print('✅ Gitea Wiki Remote synchronisation complete.')
    else:
        print('   No changes detected in Wiki. Push skipped.')

    print('🎉 Wiki Deployment and Synchronization Pipeline finished.')

if __name__ == '__main__':
    main()
